lib/linguist/languages.yml

```diff
@@ -966,6 +966,14 @@ Cairo:
   tm_scope: source.cairo
   extensions:
   - ".cairo"
+  language_id: 620599568
```

Comment:
`script/update-ids` generated the same `language_id` as for `Cairo 0` (620599567). This is wrong, so I have just added 1, as I see it's just hashing language name so this sounds safe?

Reply:
🤔 It shouldn't.

This is where where the magic happens:

https://github.com/github-linguist/linguist/blob/39fd5e93de98de3434cb3e857c5b1972c418f8c5/script/update-ids#L8-L10

```ruby
irb(main):004:0> require 'digest'
=> true
irb(main):005:0> language = 'Cairo 0'
irb(main):006:0> Digest::SHA256.hexdigest(language).to_i(16) % (2**30 - 1)
=> 891399890
irb(main):007:0> language = 'Cairo'
irb(main):008:0> Digest::SHA256.hexdigest(language).to_i(16) % (2**30 - 1)
=> 620599567
irb(main):009:0> 
```

It however won't update an entry that already has an ID - we don't want the IDs in GitHub's database to change. As such, it's better to create a whole new `Cairo 0` entry with new ID and keep the current `Cairo` with the current ID.

I recommend adding `group: Cairo` line to the new `Cairo 0` entry too so it appears to the user that it's all Cairo.

Reply:
> As such, it's better to create a whole new `Cairo 0` entry with new ID and keep the current `Cairo` with the current ID.

Question: Will existing files be reclassified when a linguist change is released, which reclassifies files from one language ID to another? 

> I recommend adding `group: Cairo` line to the new `Cairo 0` entry too so it appears to the user that it's all Cairo.

I'm not sure about this. For Starknet developers (majority of Cairo use) it's pretty important to know whether a codebase it Cairo 0 or Cairo 1.0+. Cairo 0 is not deployable any more on all official networks.

Given this context, would you still recommend grouping these?

Reply:
> Question: Will existing files be reclassified when a linguist change is released, which reclassifies files from one language ID to another?

No. A repo will only be reclassified when reanalysed which only happens on pushes to the repo.

It's common for very old repos to show the old name until a new push is made, which is kinda right because that was what the language was called at the time of the last push. You'll see this for the variations of VBNet.

Reusing a name with a different ID really confuses things and leads to a lot of support issues because of caching. This is also complicated by Search which uses a different library which feeds off linguist, hence the suggestion of adding a new language for the old name with a new name. 

> I'm not sure about this. For Starknet developers (majority of Cairo use) it's pretty important to know whether a codebase it Cairo 0 or Cairo 1.0+. Cairo 0 is not deployable any more on all official networks.

The grouping only affects what is shown in the sidebar and as the primary language for a repo based on bytes of code. Individual files will be identified by the individual names. Additionally, given the vastly different syntax, wouldn't this be easily recognisable by users?

Reply:
> Additionally, given the vastly different syntax, wouldn't this be easily recognisable by users?

Yes, they can.

---

Done